### DISCUSSION TITLE
Comprehensive side effect audit logging

### DISCUSSION THREAD

**Discussion by Emyrk** [coder/coder#5419] [RFCs]
# Problem

Currently audit logging is initiated per route. This means that each route has at most **1** audit event created. _However_, single http request can cause multiple changes to multiple entities. This is currently not being tracked.

### Create workspace

Create workspace has multiple db changes; `workspace`, `parameters`, `provisioner_job`, and `workspace_build`. Only `workspace` is logged even though 4 different db changes are made.

These tables can be updated in other routes and generate different audit log events (eg `workspace_build`). But because it is a side effect in the `create workspace` route, it is not logged the same as it is in `update workspace`.

# Question

Should we design/implement a way to audit log side effects?

#### Create workspace

An audit log for the create workspace should be made, as that is the **user action**. Side effects of the provisioner job and workspace build should also be logged so that they can be tracked in the context of the original user action.

Now a `provisioner` can be searched to see what jobs it ran and see the original context for each job. Without the linking and side effect audit logs, these jobs are impossible to track the origin because only the `workspace` diff and id are logged.

# Proposal

## Linked audit logs (parent -> child or primary -> secondary)

Allow creating linked audit logs with an optional `parent_id` field. Setting a `parent_id` of an audit log allows relating to a primary action. In the UI only parent logs should be shown. Children logs should be hidden under some sort of "toggle" to see all side effects of a primary action.

We should not allow infinite nesting, and only allow 1 level of "children" logs. We do not want to build full tracing here.

## Linked audit log table

A new audit log table for these secondary audit logs would be created.

```
audit_diffs
- primary_log (id)
- resource_id
- resource_type (string)
- diff
- succeeded true/false
```

## Other ideas?

You got any?

**Comment by Kira-Pilot** [comment]
So glad you brought this up! It's definitely a pain point. Some color:
1. As you may be aware, we do [actually log workspace build jobs](https://github.com/coder/coder/blob/main/coderd/provisionerdserver/provisionerdserver.go#L530), but you're absolutely right: they're not really associated with the `workspace` resource in any way. As a result, we are missing some information that would be nice to have, e.g. [the IP address](https://github.com/coder/coder/issues/5268) associated with the request.
2. I'm running into this issue right now as I try to audit `group_members`, which exist in a table separate from `group`. You can see some discussion in the [open PR ](https://github.com/coder/coder/pull/5357/files#r1043635183)where we consider the pros and cons of adding another auditable resource.

So pain points from my end are scaling (how many auditable resources do we really want) and UX (how many audit log lines do we really want; how can we make associations between log items clear; how do we handle filtering).

**Reply by Emyrk** [reply]
Thank you for the extra context links. 

A thought on secondary audit events is that many can be implemented on the persistence layer itself. So every change to the db will automatically drop these logs.

-----

You bring up even more points of "secondary events" such as failed jobs. This would give the provisioner a way to drop an audit logs as a failed event relating to the primary user action.

This does bring up a question though. Should the provisioner log show in the UI as it's own row? Or underneath the `Create Workspac` audit log row in some toggle?

**Reply by Kira-Pilot** [reply]
I like the idea of toggling because I actually think it might keep the UI simpler. What I would like to avoid is having to put in a bunch of filtering so users can make sense of the view - as @spikecurtis mentioned below, most filtering is probably going to be done in Splunk. IDK though - warrants more thought.

**Comment by spikecurtis** [comment]
In the RFC, we set this [as a requirement](https://www.notion.so/coderhq/v2-Enterprise-Audit-Logging-a6358d50a453413eae6441e689f023ac#3268f1c06a7e43e1af738fb3b9f84ef5), and the proposed solution was to generate a RequestID that would tie together multiple logs if the request resulted in multiple actions.

**Reply by Emyrk** [reply]
`request_id` is a field on audit logs, so we already support linking multiple logs together. :+1:

A difference between just a request ID and the proposed is that the proposed has a `primary` and `secondary` to distinguish the user action vs side effects.

I do think this audit log line is sufficient for the action taken by the user.

![Screenshot from 2022-12-14 12-28-53](https://user-images.githubusercontent.com/5446298/207679204-6ee4ee00-6090-4e19-8c69-bb66666a4623.png)

It would be excessive to also add more lines for:
- Workspace build inserted to database
- Provisioner job inserted to database
- etc

The audit log UI would be very cluttered. So the idea of a `secondary` event allows nesting. TBD if the audit logs page should allow sorting/showing by secondary events. Maybe there is a view that does not collapse them?

**Reply by spikecurtis** [reply]
I'd be interested in customer feedback on this point, but my working assumption is that we shouldn't go 🍌 🍌 🍌  making our UI a best-in-class experience because what most customers want to do is push our audit logs into Splunk (or similar) where they can see the logs in context with logs from other systems.

**Reply by Kira-Pilot** [reply]
I would like an easy way to furnish secondary audit logs with additional detail, i.e. I would like to be able to know that a workspace build was initiated from a machine with a certain IP, or I would like to know that group members were added to a group with a specific name. Right now, our solution for getting that additional detail is to a) just not have it available, as in the case with the IP address or b) query for it, often in the route itself, and pass it along in a field called `AdditionalFields`, which is just json. Example [here](https://github.com/coder/coder/blob/main/coderd/provisionerdserver/provisionerdserver.go#L543).

It seems worth it to me to explore this issue from an architectural perspective, even if we do not make immediate changes in the UI. It would be a bummer to audit 20 different resources, have a bunch of `AdditionalField` info everywhere, and then realized down the line it's going to be a mess to change.

**Reply by Emyrk** [reply]
It does seem like we might have to store the `request_id` or whatever ID is linking the audit logs in the db somewhere.

Originally I was thinking it can all go on the `ctx`, but @Kira-Pilot brings up examples where that is not the case.

**Reply by spikecurtis** [reply]
I mean, we should already be storing it on the Audit Logs, but in the case that a request writes something to the database which then gets picked up by some background process (e.g. builds) then yes, the job that gets picked up would need the request ID.

**Reply by Emyrk** [reply]
I do like the idea of `request_id` since it allows linking audit logs to other things too.

I still think there is value in `primary` and `secondary`, not just for a UI purpose. The `primary` action should be something a user can do in the product and makes intuitive sense. `secondary` audit logs should be granular and might only make sense to a more technical user.

**Reply by bpmct** [reply]
Have we done any investigation into how other products that often involve linked, async requests handle their audit logs (e.g. CI vendors (e.g. circleci), git providers (e.g. gitlab), or identify-based access tools (e.g. teleport)? 

I'm personally curious if there are any patterns in how they link/relate logs, mostly to ensure that we don't do something bespoke that may be confusing for customers who are used to consuming logs from these other platforms (via the UI, consuming via the API, or ingesting into Splunk/StackDriver)

**Reply by Emyrk** [reply]
@bpmct I have not, but this is essentially tracing which is rather normal. The only _flair_ is how we present it on our UI

**Comment by Emyrk** [comment]
Another reason for this is we log `Login` events for v1, but signing-out is logged as an `api-key` `delete` action. It would be better to be consistent and have a `logout` audit log, but the actual action deletes 1 or more api keys.

If we can link audit log actions, then we could have a `login` and `logout` event with the api key creation/deletes logged separately.
